timemachine/fe/free_energy.py

```diff
@@ -502,3 +496,110 @@ def run_sequential_sims_given_initial_states(
         prev_state = state
 
     return np.array(u_kln_by_component_by_lambda), stored_frames, stored_boxes
+
+
+def sample_state(
+    initial_state: InitialState, md_params: MDParams, temperature: float
+) -> EnergyDecomposedState[StoredArrays]:
+    frames, boxes = sample(initial_state, md_params, max_buffer_frames=1000)
+    bound_impls = [p.bound_impl(np.float32) for p in initial_state.potentials]
+    assert initial_state.barostat is None or initial_state.barostat.temperature == temperature
+    batch_u_fns = get_batch_u_fns(bound_impls, temperature)
+    return EnergyDecomposedState(frames, boxes, batch_u_fns)
+
+
+def compute_bar_error(s1: EnergyDecomposedState, s2: EnergyDecomposedState) -> float:
+    u_kln = compute_energy_decomposed_u_kln([s1, s2]).sum(axis=0)  # sum over components
+    assert u_kln.shape[:2] == (2, 2)
+    w_f = u_kln[1, 0] - u_kln[0, 0]
+    w_r = u_kln[0, 1] - u_kln[1, 1]
+    _, df_err = bar_with_bootstrapped_uncertainty(w_f, w_r)
+    return df_err
+
+
+def run_sims_with_greedy_bisection(
+    initial_lambdas: Sequence[float],
+    make_initial_state: Callable[[float], InitialState],
+    md_params: MDParams,
+    n_bisections: int,
+    temperature: float,
+    verbose: bool = True,
+) -> Tuple[List[Tuple[List[InitialState], NDArray]], List[StoredArrays], List[NDArray]]:
+    r"""Starting from a specified lambda schedule, successively bisect the lambda interval between the pair of states
+    with the largest BAR :math:`\Delta G` error and sample the new state with MD.
+
+    Parameters
+    ----------
+    initial_lambdas: sequence of float, length >= 2, monotonically increasing
+        Initial protocol; starting point for bisection.
+
+    make_initial_state: callable
+        Function returning an InitialState (i.e., starting point for MD) given lambda
+
+    md_params: MDParams
+        Parameters used to simulate new states
+
+    n_bisections: int
+        Number of bisection steps to perform
+
+    temperature: float
+        Temperature in K
+
+    verbose: bool
+        Whether to print diagnostic information
+
+    Returns
+    -------
+    results: list of tuples of (list of InitialState, NDArray)
+        Results from each iteration of bisction. The first element of the nth tuple is the list of InitialState objects
+        specifying the protocol after the nth iteration of bisection; the second element is a (L-1, 2, 2, F) array of
```

Comment:
Confirm if `(L-1, 2, 2, F)` is the correct shape? (Is there also an axis with size # of energy components?)

Reply:
Fixed in https://github.com/proteneer/timemachine/pull/959/commits/6b57d661ee4b9559ecd7691665297bfd771e264f